engine/enginefuzz_test.go

```diff
@@ -298,3 +298,248 @@ func validateTestCases(t *testing.T, cases []*testCase) {
+		count2 := bucket2[0] + bucket2[1] + bucket2[2]
+
+		if count1 == 0 || count2 == 0 {
+			return
+		}
+
+		load := fmt.Sprintf(`load 2m
+			http_request_duration_seconds{pod="nginx-1"} {{schema:%d count:%d sum:%.2f buckets:[%d %d]}}+{{schema:%d count:%d sum:%.2f buckets:[%d %d %d]}}x20 
+			http_request_duration_seconds{pod="nginx-2"} {{schema:%d count:%d sum:%.2f buckets:[%d]}}+{{schema:%d count:%d sum:%.2f buckets:[%d %d %d]}}x20`,
+			schema1, count1-bucket1[2], sum1, bucket1[0], bucket1[1],
+			schema1, count1, sum1, bucket1[0], bucket1[1], bucket1[2],
+
+			schema2, count2-bucket2[1]-bucket2[2], sum2, bucket2[0],
+			schema2, count2, sum2, bucket2[0], bucket2[1], bucket2[2],
+		)
+
+		opts := promql.EngineOpts{
+			Timeout:              1 * time.Hour,
+			MaxSamples:           1e10,
+			EnableNegativeOffset: true,
+			EnableAtModifier:     true,
+			EnablePerStepStats:   true,
+		}
+
+		qOpts := promql.NewPrometheusQueryOpts(true, 0)
+		storage := promqltest.LoadedStorage(t, load)
+		defer storage.Close()
+
+		queryTime := time.Unix(int64(ts), 0)
+
+		seriesSet, err := getSeries(context.Background(), storage, "http_request_duration_seconds")
+		require.NoError(t, err)
+
+		psOpts := []promqlsmith.Option{
+			promqlsmith.WithEnableOffset(true),
+			promqlsmith.WithEnableAtModifier(true),
+			promqlsmith.WithEnabledAggrs([]parser.ItemType{
+				parser.SUM, parser.MIN, parser.MAX, parser.AVG, parser.GROUP, parser.COUNT, parser.COUNT_VALUES, parser.QUANTILE,
+			}),
+		}
+
+		ps := promqlsmith.New(rnd, seriesSet, psOpts...)
+		newEngine := engine.New(engine.Opts{EngineOpts: opts, EnableAnalysis: true})
+		oldEngine := promql.NewEngine(opts)
+
+		cases := make([]*testCase, 0, testRuns)
+		for i := 0; i < testRuns; i++ {
+			var (
+				q1    promql.Query
+				query string
+			)
+
+			for {
+				expr := ps.WalkInstantQuery()
+				query = expr.Pretty(0)
+
+				q1, err = newEngine.NewInstantQuery(context.Background(), storage, qOpts, query, queryTime)
+				if engine.IsUnimplemented(err) || errors.As(err, &parser.ParseErrors{}) {
+					continue
+				} else {
+					break
+				}
+			}
+			testutil.Ok(t, err)
+			newResult := q1.Exec(context.Background())
+
+			newStats := q1.Stats()
+			stats.NewQueryStats(newStats)
+
+			q2, err := oldEngine.NewInstantQuery(context.Background(), storage, qOpts, query, queryTime)
+			testutil.Ok(t, err)
+			oldResult := q2.Exec(context.Background())
+
+			oldStats := q2.Stats()
+			stats.NewQueryStats(oldStats)
+
+			cases = append(cases, &testCase{
+				query:           query,
+				newRes:          newResult,
+				newStats:        newStats,
+				oldRes:          oldResult,
+				oldStats:        oldStats,
+				loads:           []string{load},
+				start:           queryTime,
+				end:             queryTime,
+				validateSamples: false,
+			})
+		}
+
+		if len(cases) > 0 {
+			validateTestCases(t, cases)
+		}
+	})
+}
+
+func FuzzNativeHistogramOnRangeQuery(f *testing.F) {
```

Comment:
I suggest just joining these two together with some extra parameters because a lot of stuff is copied/pasted.